core/chains/evm/logpoller/orm.go

```diff
@@ -708,35 +711,22 @@ func (o *DbORM) blocksRangeAfterTimestamp(after time.Time, confs Confirmations,
 	if len(blocks) != 2 {
 		return 0, 0, nil
 	}
+	// To stay consistent between requesting finalized and non-finalized logs, we always return min(block_number) as the lower range.
+	// Finalized setting impacts only the upper range. So we either pick max(lastFinalizedBlock) or max(block_number) - confs
+	if confs == Finalized {
+		return blocks[0].BlockNumber, blocks[1].LastFinalizedBlockNumber, nil
+	}
 	return blocks[0].BlockNumber, blocks[1].BlockNumber - int64(confs), nil
 }
 
-func (o *DbORM) SelectLogsUntilBlockHashDataWordGreaterThan(address common.Address, eventSig common.Hash, wordIndex int, wordValueMin common.Hash, untilBlockHash common.Hash, qopts ...pg.QOpt) ([]Log, error) {
-	var logs []Log
-	q := o.q.WithOpts(qopts...)
-	err := q.Transaction(func(tx pg.Queryer) error {
-		// We want to mimic the behaviour of the ETH RPC which errors if blockhash not found.
-		var block LogPollerBlock
-		if err := tx.Get(&block,
-			`SELECT * FROM evm.log_poller_blocks 
-					WHERE evm_chain_id = $1 AND block_hash = $2`, utils.NewBig(o.chainID), untilBlockHash); err != nil {
-			return err
-		}
-		return q.Select(&logs,
-			`SELECT * FROM evm.logs 
-			WHERE evm_chain_id = $1
-			AND address = $2 AND event_sig = $3
-			AND substring(data from 32*$4+1 for 32) >= $5
-			AND block_number <= $6 
-			ORDER BY (block_number, log_index)`, utils.NewBig(o.chainID), address, eventSig.Bytes(), wordIndex, wordValueMin.Bytes(), block.BlockNumber)
-	})
-	if err != nil {
-		return nil, err
+func nestedBlockNumberQuery(confs Confirmations) string {
+	if confs == Finalized {
+		return `
+				(SELECT COALESCE(last_finalized_block_number, 0) 
+				FROM evm.log_poller_blocks 
+				WHERE evm_chain_id = :evm_chain_id 
+				ORDER BY block_number DESC LIMIT 1)`
 	}
-	return logs, nil
-}
-
-func nestedBlockNumberQuery() string {
 	return `
 			(SELECT COALESCE(block_number, 0) 
 			FROM evm.log_poller_blocks 	
```

Comment:
I missed this in the last PR, but just noticed we're subtracting `- :confs` from `COALESCE(block_number, 0)` which looks rather odd since it could result in a negative number.  I guess that's probably okay because `block_number` can only be NULL for very old blocks (due to a previous migration IIRC)?

Reply:
Actually, it looks like there is a `NOT NULL` constraint on `block_number`.  I think we should be able to just replace it with this?

```
(SELECT block_number - :confs FROM evm.log_poller_blocks 	
			WHERE evm_chain_id = :evm_chain_id 
			ORDER BY block_number DESC LIMIT 1)`
```
Maybe @connorwstein can weigh in here, as it looks like the original `COALESCE` goes all the way back to this commit. Was there a purpose to this originally or did it just end up there somehow accidentally?
https://github.com/smartcontractkit/chainlink/commit/92d6601ab6ddc1b20cbde2a6cfde4d1d82851907#diff-edecd64e94d520ea2ed55dacabae7c260f4c0bd06245fe003bc255fdd1c6b28bR65-R77

Reply:
It looks odd, but technically works exactly the same as before

We started with
```sql
 AND (block_number + $4) <= (SELECT COALESCE(block_number, 0) FROM evm.log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)
```

and then moved to
```sql 
            AND block_number <= (SELECT COALESCE(block_number, 0) FROM evm.log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1) - $4
```

These two queries are equal. I agree it might look a bit odd, but this query is nested so it doesn't change much.

Reply:
`COALESCE(block_number, 0)` works as an additional guard. Null can be returned when table or no records can be found for a specific `evm_chain_id`. As this query is not executed alone (it's always inserted as a nested query) I would keep coalescing.

Reply:
`block_number < NULL` will just evaluate to null even if the lefthand side isn't NULL which is the same as FALSE if used in a WHERE clause.  So I don't think having a negative number vs NULL makes any difference for the queries we have.  But I guess we could leave it as a guard in case someone tries to use it for something more fancy.

Aesthetically, it might look a bit better if we changed it to `SELECT COALESCE(block_number - :confs, 0)`, unless that somehow affects performance. I guess it's not a big deal whatever we do here, since these inequalities will behave the same with 0 or negative anyway--I'm fine with leaving it.

Reply:
@reductionista I will do a small test, I believe we can drop `coalesce` and simplify the query as much as possible as long as it doesn't affect performance. At this moment it's making it only confusing

Reply:
Originally I was thinking that having `(SELECT COALESCE(block_number, 0) FROM log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)` return 0 as the latest block on an empty block table made intuitive sense for confirmations filters. That is if you wanted 6 confs on top of X and the latest block is 0, you don't have enough.  Using greatest instead seems fine to me

Reply:
I like the latest change @mateusz-sekara made:
```
(SELECT greatest(block_number - :confs, 0) 
			FROM evm.log_poller_blocks 					FROM evm.log_poller_blocks 		
```
Looks like `greatest(NULL, 0)` returns 0, which is ideal.  Guards against any potential weird use of this sub query, as well as handling the additional case where `block_number < :confs` elegantly.

Reply:
@connorwstein Neither `coalesce` nor `greatest` supports a case in which the block table is empty
```sql
chainlink=# select coalesce(block_number - 10, 0) from evm.log_poller_blocks where evm_chain_id = 10;;
 coalesce
----------
(0 rows)

chainlink=# select greatest(block_number - 10, 0) from evm.log_poller_blocks where evm_chain_id = 10;;
 greatest
----------
(0 rows)
```

There is a difference in Postgres between the empty result set and the null value returned from the column. `NULL`s can't be returned from `evm.log_poller_block` because the db constraint enforces block_number to be positive value, so right now, `coalesce` is just redundant.

Reply:
The only solution to always return right now is to have sth like this, not sure if worth the effort
```sql
SELECT GREATEST(block_number - 10, 0) 
FROM evm.log_poller_blocks 
WHERE evm_chain_id = 10

UNION ALL

SELECT 0
WHERE NOT EXISTS (
    SELECT 1 FROM evm.log_poller_blocks WHERE evm_chain_id = 10
);
```